### PR TITLE
Check actual resources consumption in work_queue_hungry

### PR DESCRIPTION
**Pull request by tphung3**
The check for actual resources consumption can be improved to avoid the problem of tasks being submitted in a loop while resources are fragmented? For example, a more comprehensive check that matches specific available workers to appropriate tasks. What do you think?

### CODE REVIEW

**Comment by dthain** [comment]
Yes, this looks like the right idea.
Make sure to test this with a simple application to be sure all the corner cases are handled.
For example, what happens if there are no workers connected yet?

**Comment by btovar** [comment]
One thing that worries me is the initial state of hungry and accepting new workers. Say my application does something like:

```sh
while queue not empty:
    if queue is hungry:
       submit some tasks
    wq wait
```

At the start, there are no tasks and no workers connected, so queue hungry is false. Since no task is ever submitted, wq wait returns immediately and no new workers connect. With this, hungry is always false.

**Comment by dthain** [comment]
@tphung3 please push your latest changes, so we can wrap this one up.

**Comment by dthain** [comment]
Ah, I think this may be too complicated -- you don't want to block inside this call, because nothing will happen while you are sleeping.

Better to just return true in any unusual case, which will encourage the caller to compose a job and call `work_queue_wait`

**Comment by dthain** [comment]
Let me be more specific:

- Count up the number and resources of jobs waiting, as you are doing now.

- Return true if the sum of those resources is less than the sum of the resources of the workers.  (We want enough tasks to feed all available workers.)

- Return true if the number of waiting jobs is less than, say, 10.  (We want a minimum number of tasks available regardless of workers.)

Does that make sense?

**Comment by tphung3** [comment]
Yes that makes sense. I agree with point 1 and 3 but point 2 can be improved. Assume that we have a lot of small-sized tasks that are ready and workers can only take **some** of them, but the total required resources are higher than the amount of resources workers have, so in this case we actually are not submitting anything, as `work_queue_hungry` always return false. So instead of summing up the required resources, maybe we can check the ready tasks individually?

**Comment by dthain** [comment]
I don't quite follow.  Can you give an example?

**Comment by tphung3** [comment]
Let's say all workers have 10 cores and 1GB of memory and disk available. We also have 100 tasks, and each requires 1 core and 100 MB of memory and disk. If we sum up all required resources of tasks, then WQ hungry will return false since all 100 tasks amount to 100 cores and 10GB of memory and disk. But if we check each task one at a time, then WQ hungry will return true as the workers can take in a small number of tasks (10 in this case).

**Comment by tphung3** [comment]
How likely do you think it can actually happen?

**Comment by dthain** [comment]
Ah, I see what you mean.

So, from the big picture, we have to balance several things:

- We want to have a sufficient backlog of idle jobs in the queue, so that WQ can dispatch new tasks quickly when old ones complete.  But, we don't want to materialize an unnecessary number of idle jobs, otherwise the master's time/memory will be occupied managing the queue.
- The time/complexity of work_queue should be low, because it is going to get called about once for every submission.  We don't want to re-implement the whole scheduling algorithm here, which takes into account data locality, network performance, special capabilities, etc.  If we have to simplify, then work_queue_hungrey should err on the side of true, rather than false.

You are right that simply counting up the resources is going to result in some inaccuracy, because the tasks may not pack into the workers perfectly.

In your example, the tasks don't nicely pack into the workers, because the tasks don't consume much memory.  What will happen is that WQ will assign tasks to workers until the critical resource (cores) is exhausted.  So, you will get 10 tasks per worker, which is the desired outcome.

Thus, we want work_queue_hungry to keep returning true until the critical resource (in this case, cores) is exhausted.

**Comment by tphung3** [comment]
The field `q->stats->tasks_waiting` seems to fail at tracking the number of ready tasks so I used 'q->stats->submitted - q->stats->dispatched' instead.

File: work_queue/src/work_queue.c [hunk 1]
```diff
@@ -6184,17 +6184,73 @@ struct work_queue_task *work_queue_wait_internal(struct work_queue *q, int timeo
...
-	if(q->stats->tasks_dispatched < 100)
-		return MAX(100 - ready, 0);
+	//check if queue is initialized
+	//return false if not
+	if (q == NULL){
+		return 0;
+	}
 
-	//BUG: fix this so that it actually looks at the number of cores available.
+	//check if queue's statistics are initialized
```
Comment (dthain) [inline work_queue/src/work_queue.c:6193]: Ah, I see the problem here.  `q->stats` is used internally to track some stats incrementally, but doesn't track those that are expensive to compute.  This is done in `work_queue_get_stats` which takes some things from `q->stats` and then adds the computed items.

So, what you want to do is this:
```
work_queue_stats s;
work_queue_get_stats(q,&s);
```

Then everything in `s` (like `tasks_waiting` will be up to date and can be used normally.

File: work_queue/src/work_queue.c [hunk 2]
```diff
@@ -6184,17 +6184,73 @@ struct work_queue_task *work_queue_wait_internal(struct work_queue *q, int timeo
...
+		if (q->stats->tasks_submitted - q->stats->tasks_dispatched < 10){
+			return 4;
+		}
+		return 0;
+	}
+	
+	//if number of ready tasks is less than 10, return true for more tasks in queue 
+	//ready = submitted - dispatched
+	//10 is chosen to be the default number of ready tasks in queue to keep queue efficient
+	if (q->stats->tasks_submitted - q->stats->tasks_dispatched < 10){
```
Comment (dthain) [inline work_queue/src/work_queue.c:6213]: As above, use `s.tasks_waiting`

**Review by dthain — CHANGES_REQUESTED** [review]
Almost there: see comment about `q->stats`

**Comment by dthain** [comment]
Yes, looks good!
One more thing, add gpus as a fourth resource.
(I just fixed this up in a different PR.)

**Review by dthain — APPROVED** [review]

**Comment by dthain** [comment]
@tphung3 go ahead and close #2497!
